ROCpd support [Part 1]

adjordje-amd opened the pull request:
Add ROCpd support:
 - cpu_frequency
 - amd_smi
 - sampling

# rocprofiler-systems Pull Request

## Related Issue
<!-- Please link to the external GitHub issue(s) that this PR addresses. 
  If providing a JIRA ticket, please don't include an internal link -->
- [ ] Closes #<issue number>

## What type of PR is this? (check all that apply)

- [ ] Bug Fix
- [ ] Cherry Pick
- [ ] Continuous Integration
- [ ] Documentation Update
- [x] Feature
- [ ] Optimization
- [ ] Refactor
- [ ] Other (please specify)

## Technical Details
<!-- Please explain the changes. -->

## Have you added or updated tests to validate functionality?

- [ ] Yes
- [x] No - does not apply to this PR

## Added / Updated documentation?

- [ ] Yes
- [x] No - does not apply to this PR

## Have you updated CHANGELOG?
<!-- Needed for Release updates for a ROCm release. -->
- [ ] Yes
- [x] No - does not apply to this PR

adjordje-amd committed:
Add rocpd support for
 - cpu_frequency
 - amd_smi
 - sampling

dgaliffiAMD committed:
Adding sqlite-dev to our ci docker images (#280)

sputhala-amd committed:
Make CMakeFile fixes to align with rocDecode and rocJpeg changes (#281)

darren-amd committed:
Fix ROCtx event ranges in trace output (#278)

* Fix marker api traces

* Remove space

* Formatting change

* Small change

* Update Changelog

* Add period to changelog

* Update source/lib/rocprof-sys/library/rocprofiler-sdk.cpp

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

* Fix roctx tests

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

sputhala-amd committed:
Update gotcha submodule from timemory (#277)

* Update gotcha submodule from timemory

* Fix build failure and add copilot suggestions

* Fix formatting errors

adjordje-amd committed:
Fix compile errors, copilot comments

adjordje-amd committed:
Fix shadowed variables

adjordje-amd committed:
Fix agent include when ROCM is not used

dgaliffiAMD committed:
Remove an undefined submodule from the GOTCHA submodule (#285)

A recursive submodule update, `git submodule uppdate --recursive
--init`,
would fail due to an improperly defined and empty submodule in GOTCHA.

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

marantic-amd committed:
Refactor agent manager to work when ROCM is disabled

mradosav-amd commented on source/lib/rocprof-sys/library/components/comm_data.cpp:
```diff
@@ -74,6 +78,119 @@ write_perfetto_counter_track(uint64_t _val)
+        BLOCK, EXPRESSION, 0, 0);
+#endif
+#if defined(ROCPROFSYS_USE_RCCL)
+    data_processor.insert_pmc_description(
```
```suggestion
#if defined(ROCPROFSYS_USE_MPI)
    data_processor.insert_pmc_description(
        ni.id, getpid(), base_id, TARGET_ARCH, EVENT_CODE, INSTANCE_ID,
        mpi_send::label, "Tracks MPI Send communication data sizes",
        trait::name<category::mpi>::description, LONG_DESCRIPTION, COMPONENT, MSG, "ABS",
        BLOCK, EXPRESSION, 0, 0);
    data_processor.insert_pmc_description(
        ni.id, getpid(), base_id, TARGET_ARCH, EVENT_CODE, INSTANCE_ID,
        mpi_recv::label, "Tracks MPI Recv communication data sizes",
        trait::name<category::mpi>::description, LONG_DESCRIPTION, COMPONENT, MSG, "ABS",
        BLOCK, EXPRESSION, 0, 0);
#endif
```

habajpai-amd committed:
Add missing <cstring> include for C string functions in RCCL tests (#282)

* Fix: Add missing <string.h> include for C string functions in RCCL tests

* Update examples/rccl/rccl-tests/src/common.h

Yes, confirmed—<cstring> alone works in my environment. Updated the PR

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

* clang-format

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

ajanicijamd committed:
Add a CMake Presets file (#44)

Adds the following presets: 

- `ci` - to match the common CI settings - including tests and asserts
- `debug` - True debug build - include building tests
- `debug-optimized` - include building tests
- `release` - To match the "build-release` script - no tests.

The default build folder will be `${sourceDir}/build/<preset>`.

---------

Co-authored-by: David Galiffi <David.Galiffi@amd.com>

marantic-amd committed:
Fix compile errors and typos

marantic-amd committed:
Fix compile errors and typos

adjordje-amd committed:
Move agent and agent_manager from rocpd namespace

adjordje-amd committed:
Merge remote-tracking branch 'refs/remotes/origin/rocpd-implementaion-part-1' into rocpd-implementaion-part-1

adjordje-amd committed:
Replace std::filepath with a timemory filepath impl

adjordje-amd committed:
Fix formatting

adjordje-amd committed:
Fix undefined mpi_send

mradosav-amd committed:
Fix undefined mpi_recv

mradosav-amd committed:
Move use_rocpd guard in comm_data

adjordje-amd committed:
Benchmark support same category from different threads

adjordje-amd committed:
Merge branch 'amd-staging' into rocpd-implementaion-part-1

AlexandarDjordjevic commented on source/lib/core/rocpd/data_storage/database.hpp:
```diff
@@ -0,0 +1,187 @@
+            int             position   = 1;
+            auto            bind_value = [&](auto _value) {
+                using T = decltype(_value);
+                if constexpr(std::is_same_v<T, int32_t> || std::is_same_v<T, uint32_t>)
```
Create a template specialization to bind the value and replace if/else

AlexandarDjordjevic commented on source/lib/common/traits.hpp:
```diff
@@ -0,0 +1,91 @@
+{
+
+template <typename T>
+struct is_string_literal : std::false_type
```
improve traits with std::decay

adjordje-amd committed:
Code improvement

adjordje-amd commented on source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -0,0 +1,170 @@
+#include <fstream>
+#include <regex>
+#include <timemory/environment/types.hpp>
+#include <unistd.h>
```
```suggestion

#include <config.hpp>
#include <fstream>
#include <regex>
#include <timemory/environment/types.hpp>
#include <timemory/utility/filepath.hpp>
#include <unistd.h>
```

adjordje-amd committed:
Update source/lib/core/rocpd/data_storage/database.cpp

adjordje-amd committed:
Add ROCPROFSYS_ENABLE_BENCHMARK option

dgaliffiAMD committed:
Update CHANGELOGS.md

dgaliffiAMD commented on examples/jpegdecode/CMakeLists.txt:
It seems like this file can be reverted from the PR. Maybe it was pulled in from a merge.

dgaliffiAMD commented on source/lib/core/benchmark/benchmark.hpp:
```diff
@@ -0,0 +1,353 @@
+#include "core/benchmark/category.hpp"
+#include "core/debug.hpp"
+
+namespace rocprofiler
```
```suggestion
namespace rocprofsys
```
Changing namespace to match others in project.

dgaliffiAMD commented on source/lib/core/benchmark/category.hpp:
```diff
@@ -0,0 +1,68 @@
+
+#include <string_view>
+
+namespace rocprofiler
```
```suggestion
namespace rocprofsys
```
Changing namespace to match others in project.

dgaliffiAMD commented on source/lib/core/benchmark/category.hpp:
```diff
@@ -0,0 +1,68 @@
+}
+
+}  // namespace benchmark
+}  // namespace rocprofiler
```
```suggestion
}  // namespace rocprofsys
```

dgaliffiAMD commented on source/lib/core/benchmark/benchmark.hpp:
```diff
@@ -0,0 +1,353 @@
+}
+
+}  // namespace benchmark
+}  // namespace rocprofiler
```
```suggestion
}  // namespace rocprofsys
```

dgaliffiAMD commented on source/lib/core/CMakeLists.txt:
```diff
@@ -10,6 +10,7 @@ set(core_sources
     ${CMAKE_CURRENT_LIST_DIR}/categories.cpp
     ${CMAKE_CURRENT_LIST_DIR}/config.cpp
     ${CMAKE_CURRENT_LIST_DIR}/constraint.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/cpu.cpp
```
```
set(core_sources
    ${CMAKE_CURRENT_LIST_DIR}/agent_manager.cpp
    ${CMAKE_CURRENT_LIST_DIR}/amd_smi.cpp
    ${CMAKE_CURRENT_LIST_DIR}/argparse.cpp
    ${CMAKE_CURRENT_LIST_DIR}/categories.cpp
    ${CMAKE_CURRENT_LIST_DIR}/config.cpp
    ${CMAKE_CURRENT_LIST_DIR}/constraint.cpp
    ${CMAKE_CURRENT_LIST_DIR}/cpu.cpp
    ${CMAKE_CURRENT_LIST_DIR}/debug.cpp
    ${CMAKE_CURRENT_LIST_DIR}/dynamic_library.cpp
    ${CMAKE_CURRENT_LIST_DIR}/exception.cpp
    ${CMAKE_CURRENT_LIST_DIR}/gpu.cpp
    ${CMAKE_CURRENT_LIST_DIR}/mproc.cpp
    ${CMAKE_CURRENT_LIST_DIR}/node_info.cpp
    ${CMAKE_CURRENT_LIST_DIR}/perf.cpp
    ${CMAKE_CURRENT_LIST_DIR}/perfetto.cpp
    ${CMAKE_CURRENT_LIST_DIR}/rocprofiler-sdk.cpp
    ${CMAKE_CURRENT_LIST_DIR}/state.cpp
    ${CMAKE_CURRENT_LIST_DIR}/timemory.cpp
    ${CMAKE_CURRENT_LIST_DIR}/utility.cpp
)
```
Sorting list by filename. I couldn't do it via code suggestion, but here is the list.

dgaliffiAMD commented on source/lib/core/CMakeLists.txt:
```diff
@@ -22,6 +23,8 @@ set(core_sources
+    ${CMAKE_CURRENT_LIST_DIR}/node_info.cpp
 )
 
 set(core_headers
```
```
set(core_headers
    ${CMAKE_CURRENT_LIST_DIR}/agent.hpp
    ${CMAKE_CURRENT_LIST_DIR}/agent_manager.hpp
    ${CMAKE_CURRENT_LIST_DIR}/amd_smi.hpp
    ${CMAKE_CURRENT_LIST_DIR}/argparse.hpp
    ${CMAKE_CURRENT_LIST_DIR}/categories.hpp
    ${CMAKE_CURRENT_LIST_DIR}/common.hpp
    ${CMAKE_CURRENT_LIST_DIR}/concepts.hpp
    ${CMAKE_CURRENT_LIST_DIR}/config.hpp
    ${CMAKE_CURRENT_LIST_DIR}/constraint.hpp
    ${CMAKE_CURRENT_LIST_DIR}/cpu.hpp
    ${CMAKE_CURRENT_LIST_DIR}/debug.hpp
    ${CMAKE_CURRENT_LIST_DIR}/dynamic_library.hpp
    ${CMAKE_CURRENT_LIST_DIR}/exception.hpp
    ${CMAKE_CURRENT_LIST_DIR}/gpu.hpp
    ${CMAKE_CURRENT_LIST_DIR}/locking.hpp
    ${CMAKE_CURRENT_LIST_DIR}/mpi.hpp
    ${CMAKE_CURRENT_LIST_DIR}/mproc.hpp
    ${CMAKE_CURRENT_LIST_DIR}/node_info.hpp
    ${CMAKE_CURRENT_LIST_DIR}/perf.hpp
    ${CMAKE_CURRENT_LIST_DIR}/perfetto.hpp
    ${CMAKE_CURRENT_LIST_DIR}/redirect.hpp
    ${CMAKE_CURRENT_LIST_DIR}/rocprofiler-sdk.hpp
    ${CMAKE_CURRENT_LIST_DIR}/state.hpp
    ${CMAKE_CURRENT_LIST_DIR}/timemory.hpp
    ${CMAKE_CURRENT_LIST_DIR}/utility.hpp
)
```
Sorting this too

dgaliffiAMD commented on source/lib/rocprof-sys/library/rocprofiler-sdk.cpp:
It looks like we can remove this file from thie particular PR.

dgaliffiAMD committed:
Renamed `rocpd_process_kokos_event` to `rocpd_process_kokkos_event`

dgaliffiAMD committed:
Fixed typo

Rename `rocpd_post_process_backtrace_metrix` to `rocpd_post_process_backtrace_metrics`

Signed-off-by: David Galiffi <David.Galiffi@amd.com>

dgaliffiAMD commented on source/lib/core/rocpd/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+set(rocpd_sources
+    ${CMAKE_CURRENT_LIST_DIR}/data_processor.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/json.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/data_storage/database.cpp
```
Just a question about coding convention.
Do we prefer a `CMakeLists.txt` file in each subfolder or is it okay to include all the files (of this module) in the root CMakeLists.txt.

adjordje-amd replied:
We can use a `CMakeLists.txt` file for each subfolder, as it seems that the current project structure uses this style.

dgaliffiAMD commented on source/lib/rocprof-sys/library/components/comm_data.cpp:
```diff
@@ -74,6 +74,135 @@ write_perfetto_counter_track(uint64_t _val)
+
+    get_data_processor().insert_category(category_enum_id<category::comm_data>::value,
+                                         trait::name<category::comm_data>::value);
+    get_data_processor().insert_category(category_enum_id<category::mpi>::value,
```
Do you want to guard these with `ROCPROFSYS_USE_MPI` and `ROCPROFSYS_USE_RCCL` too?

adjordje-amd committed:
Address comments

adjordje-amd committed:
Revert examples/jpegdecode/CmakeLists.txt

adjordje-amd committed:
Revert rocprofiler-sdk.cpp

adjordje-amd committed:
Merge branch 'amd-staging' into rocpd-implementaion-part-1

adjordje-amd committed:
Update CMakeLists.txt

dgaliffiAMD commented on source/lib/core/rocpd/data_processor.cpp:
```diff
@@ -0,0 +1,672 @@
+    if(it != _pmc_descriptor_map.end())
+    {
+        ROCPROFSYS_WARNING(
+            0, "Insert PMC description failed! Error: PMC descriptor already exist!\n");
```
```suggestion
            0, "Insert PMC description failed! Error: PMC descriptor (name:%s) (ID:%lu) already exist!\n",
            name, agent_id);
```
Adding a little more detail to the warning message.

dgaliffiAMD committed:
Update source/lib/core/rocpd/data_processor.cpp

dgaliffiAMD committed:
format source

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/data_views.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/marker_views.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/rocpd_indexes.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/rocpd_tables.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/rocpd_views.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.

jrmadsen commented on source/lib/core/rocpd/data_storage/schema/summary_views.sql:
rocprofiler-systems should _not_ have a local copy of this schema. Use the rocprofiler-sdk-rocpd library.
